### DISCUSSION TITLE
initializer と core との役割分担について

### DISCUSSION THREAD

**Discussion by HisashiMiura** [BRI-EES-House/heat_load_calc#235] [Ideas] [answered]
## 経緯
- 佐藤さんのVBAでの計算をもとにpythonプログラムを作成してきた。
- プログラムソースコードを、initializer と core に分割した。当初の目論見は、非住宅の計算も視野にいれており、一方で、非住宅の作業がこちらの作業に影響を極力影響を与えないように、core は負荷計算のパラメータとしての物理量としてなるべく純粋なものを残す方針とした。
- 計算量を減らすためにinitializer に壁面集約機能があった。（現在は一旦、停止している。）
- 本来であれば応答係数作成はcoreにあるべきとの議論があったため、壁面集約機能の停止とともに、応答係数作成をcoreに移した。
- 放射熱伝達率は壁の面積比率に応じて入力値から決定されるが、その計算機能はcore にあった。計算方法を厳密に見直す過程で、応答係数作成には間仕切り壁において、向こう側の空間の放射熱伝達率が必要であるが、応答係数はcoreで計算されていた。計算順序として、放射熱伝達率の計算が応答係数の計算の前にある必要があることから、応答係数の計算をcoreからinitializerに移動させた。

## 今回の提案
- これまで一旦、coreは、非住宅の計算方法の開発にも配慮して、負荷計算に純粋なものを残す方針で整理してきた。
- 一方で、非住宅特有の熱的に重い壁体（RC造など）の応答係数の計算方法も無事に組み込むことができた。
- また、事情が変わって、非住宅の計算方法にあまり配慮する必要がなくなった（スケジュールの観点から。）
- core と initializer に分割することによって、コーディングのコード量が複雑になっている。
- 前回の打ち合わせで、配列（正確にはnumpy）からjsonへの変換コードの量が膨大になっているため、core と initializer のデータ受け渡し方法を、シリアライズしたjsonではなく、データクラスを用いたいと提案した。あわせて、core と initializer に分ける必要があるのか？といった課題もでた。
- 一方で、core と initializer との役割分担を改めて考えると、省エネ基準特有の変換と、負荷計算ソフトとして汎用性のある部分とに整理することも可能ではないかと思う。例えば、外気条件の観点から見ると、省エネ基準的には地域区分を指定できるのが使い勝手が良いが、負荷計算ソフトとしては外気条件をcsvファイル等で与えることの方が望ましい。
- 現在、
a. convert：省エネ基準の入力から負荷計算に必要な入力を作成する。
b. initializer：負荷計算に必要な入力をより物理量的にピュアな変数へ変換する。
c. core：反復計算部分
となっているところを、
a. convert：省エネ基準の入力から負荷計算に必要な入力を作成する。
b. initializer：省エネ基準特有の値の処理
c. core：負荷計算に必要な入力から反復計算部分
と定義変更してもよいのではないか。
さらに、将来的には、convert と initializer の統合もありうる。
次項ではこの観点で現在initializer で行っている計算を整理する。
weather.make_weather
地域の区分から外気温湿度・日射に関する時系列データをCSVファイルとして書き出す。
→この部分は省エネ基準特有の操作。現状維持。

### initializer.make_house

1. CSVファイルから外気温湿度・日射に関する時系列データを読み込む。
→この部分は現状維持。ただし、weatherからはnumpyで取得してCSV書き出し機能はinitializerにもたせる方が整理としてはスッキリする。
2. building の整理
a. 隙間風に関して計算方法「balance_residential」を設定している。
→OK
3. spaces の整理
a. 部屋の種類のチェック（主たる居室が指定されているか、各居室が複数回指定されていないか）
→OK
b. 外気からの機械換気量の単位を、m3/h から m3/s に変換する。
→できればどちらかに統一したい。
c. 放射暖房対流比率を0.0に定めている。
→coreにもっていくべき。ただし、将来的に、パネル暖房と床暖房を併用した場合に放射暖房対流比率をどう定義していくのか？という問題は解決しないといけない。
d. 自然風利用時の換気回数を m3/s に変換する。
→できればどちらかに統一したい。
e. 熱および湿気について、家具等の熱容量・家具等と室間のコンダクタンスを、空間容積から決定する。
→学会等で十分な知見があるのであれば、initializerにある現場維持で良いが、残念ながら現時点ではあまり知見が無いため、入力値にしても入力できるユーザーはほぼ皆無である現状を考えると、coreに移動させ、デフォルト設定にするか、あるいは入力させるのかの選択肢を設けるのが良いと思う。
f. 室間換気をm3/h から m3/s に変換する。
→できればどちらかに統一したい。
4. boundaries の整理
a. sub_name = ‘’
→OK
b. 日射の有無・温度差係数が‘external_general_part’, ‘external_transparent_part’, ‘external_opaque_part’の場合のみに定義されていたものを、すべての境界で定義するように書き直す。
→この係数は直接負荷計算には使用していないが、他の計算（等価温度・透過日射）に使用しているため、他の部分の移動の有無によって影響を受ける。
c. 隣室タイプを0,1,2 間仕切り壁の場合は-1に定義しなおす。
→OK
d. 方位を、日射の有無が定義されている場合のみに書き直す。
→OK
e. 日よけ形状と日射に関する情報から透過日射量を計算する。
→coreに移す。
f. 境界の室外側の情報から相当外気温度を計算する。
→coreに移す。
g. 間仕切り壁における隣室の影響を表す係数 k_inside を定義する。
→coreに移すべきであるが、そもそも、coreの入力値として、外壁と間仕切壁を分けるべきであり、その整理がついてから整理した方が良い。
h. 在室者に対する境界の形態係数を計算する。
→coreに移す。
i. 空間の微小球に対する形態係数を計算する。
→coreに移す。
j. 放射暖房の発熱部位を表す係数flrを計算する。
→coreに移す。
k. 壁の応答係数を計算する。
→coreに移す。
5. csvで表される時系列データの作成と出力
a. 局所換気量
→OK
b. 内部発熱・発湿
→OK
c. 在室人数
→OK
d. 空調需要
→OK
e. 空調ON/OFF（空調割合）
→OK
f. 透過日射量（計算方法は既出）
→coreに移す。
g. 裏面等価温度（計算方法は既出）
→coreに移す。

## 今後の追加提案
- 壁体集約については応答係数の計算を行った後でしかできないため、一旦機能を停止している。もし応答係数をcoreに移すことができれば、壁体集約自体もcoreに移すことで復活できる可能性がある。
- ただし、現在のnumpyの使用においては、壁体の枚数が計算時間に与える影響は小さいため、特にこの作業は急ぐ必要はなく、相互放射（放射伝達率）の計算の見直しとセットで考えれば良いかと思う。

## 参考
次のEXCELファイルは、以前、佐藤さんがまとめてくれた機能表。
今回の整理をもとに修正するのが望ましい。
いずれは、表形式ではなくて、ドキュメントとして整理した後、表は削除するのが良いと思う。
 
https://github.com/BRI-EES-House/heat_load_calc/blob/master/docs/program_data_flow/%E8%A8%88%E7%AE%97%E5%86%85%E5%AE%B9%E3%81%AE%E6%95%B4%E7%90%86.xlsx

次のファイルも以前、佐藤さんがまとめて頂いた関係図。
中間ファイルから出力結果の関係性については負荷計算そのものであるため、ここで示す必要はなく、カットしても良いかと思う。
一方で、入力情報から中間ファイルへの変換はこのような図があると非常にわかりやすい。今回の整理で、この変換部分の負担は大分減ると思われるため、もっとシンプルな図になるであろう。
 
https://github.com/BRI-EES-House/heat_load_calc/blob/master/docs/program_data_flow/%E4%BD%8F%E5%AE%85%E8%B2%A0%E8%8D%B7%E8%A8%88%E7%AE%97%E3%81%AE%E3%83%87%E3%83%BC%E3%82%BF%E3%81%AE%E6%B5%81%E3%82%8C.pptx

**Comment by satoh-er** [comment, marked as answer]
@HisashiMiura 
テスト記入
